{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "name": "Candida CRISPR-Cas9 Notebook.ipynb",
      "provenance": [],
      "collapsed_sections": [],
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/breaks-software/Candida-CRISPR-Target-Notebook/blob/master/Candida_CRISPR_Cas9_Notebook.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "yaiUr_1rHFUs",
        "colab_type": "text"
      },
      "source": [
        "# Candida CRISPR-Cas9 Workbook"
      ]
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "vn8MdzWW4-hj",
        "colab_type": "text"
      },
      "source": [
        "Set up required packages:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "r6KWq7ulG_aN",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Press the Run button to set up all of the required dependencies\n",
        "from urllib import request\n",
        "from gzip import open as gzip_open\n",
        "\n",
        "# use biopython package to work with sequences\n",
        "!pip3 install biopython\n",
        "from Bio import SeqIO\n",
        "\n",
        "# install chromedriver and selenium so that we can work with the EuPaGDT site\n",
        "!apt-get update\n",
        "!apt install chromium-chromedriver\n",
        "!pip3 install selenium\n",
        "from selenium import webdriver\n",
        "from selenium.common.exceptions import NoSuchElementException\n",
        "from selenium.webdriver.support.select import Select\n",
        "from selenium.webdriver.support.ui import WebDriverWait\n",
        "from selenium.webdriver.support import expected_conditions as EC\n",
        "from selenium.webdriver.common.by import By\n",
        "\n",
        "# install the pandas package to be able to easily manipulate the data\n",
        "!pip3 install pandas\n",
        "from pandas import DataFrame, concat, read_html, options\n",
        "\n",
        "# and we need a tool to display our results in the notebook\n",
        "from IPython.display import display\n",
        "\n",
        "# we need to be able to download the results\n",
        "from google.colab import files"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "5_TJ_BDolKLY",
        "colab_type": "text"
      },
      "source": [
        "---\n",
        "## Collect the gene sequence\n",
        "In this section, we'll collect the full sequence for our target gene."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "Mr5obEnlWf_u",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Specify your target gene below, then press run:\n",
        "\n",
        "target_gene_id = \"CPAR2_404800\" #@param {type:\"string\"}"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "4QTIjsvx3aPu",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Run to download the sequence file for C. parapsilosis from the Candida Genome Database, uncompress the file for processing, then collect the sequence for our target gene\n",
        "candida_genome_database_url = 'http://www.candidagenome.org/download/sequence/C_parapsilosis_CDC317/current/C_parapsilosis_CDC317_current_orf_genomic.fasta.gz'\n",
        "\n",
        "with request.urlopen(candida_genome_database_url) as response:\n",
        "    with gzip_open(response, mode=\"rt\") as uncompressed:\n",
        "        input_seq_iterator = SeqIO.parse(uncompressed, \"fasta\")\n",
        "        try:\n",
        "          target_gene_sequence = [record for record in input_seq_iterator \\\n",
        "                                  if record.id == target_gene_id][0]\n",
        "        except IndexError:\n",
        "          raise IndexError(\"there is no sequence record for id = {}\".format(target_gene_id))      "
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "GsCsae57hXsS",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Run to review the gene sequence we found matching the specified target id:\n",
        "#@markdown Hints, \n",
        "#@markdown * the \"Description\" should be reasonable,\n",
        "#@markdown * and the [type of sequence](https://biopython.org/DIST/docs/api/Bio.Alphabet-module.html) specified at the end (e.g. SingleLetterAlphabet()) should match the type of sequence you expected.]\n",
        "print(target_gene_sequence)"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "dOn6_4of3bD2",
        "colab_type": "text"
      },
      "source": [
        "# Identify target insertion sites:\n",
        "The preference here is to insert guide RNA as early in the sequence as possible, so we want to select a target portion of the sequence.  If the sequence is long enough, target the first 2000 bases (1kb).  If the sequence is shorter than that, take the first half.  If the sequence is really short, **manually override** the sub_sequence_start and sub_sequence_length variable values below."
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "PoUJCvyQjHv1",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Change the default start position (counting starts at 0) and length for the target subsequence: \n",
        "sub_sequence_start =  0  #@param {type: \"integer\"}\n",
        "sub_sequence_length = 4017  #@param {type: \"integer\"}\n",
        "sequence_length = len(str(target_gene_sequence.seq))\n",
        "#@markdown ###  or insert your custom subsequence here (be sure to clear this if you don't want to set a custom subsequence):\n",
        "custom_sub_sequence = 'ATGGTCAAACACTTGCAATTTGTGGCGATATTGGTGGCCTTTACACTTACAGCCCTCACTCAAGCAGCTGAGATTAGCAATGTTTTCCAAAGCTTTGATAGTTTAACTTGGGAAAACGGTGCAAGCTATAGGTACAGAACCCCTTTAACTCCTAGTTGGATTGCACAATTGTCTTGGAAGATTTTAGGTTCCAATGTCAAACCAGGCGACACATTCACATTGAACATGCCATGTGTATTCAAGTTTACAACAACACAAGAGAGTATTGACTTGGACGTCGGAGACACGGTATATGCTACATGTCGTTTCGAACCAGGTGATTTAGTCGTTGCATATTCACAATTGAAATGTACTGCCAGCGATAATGTTAAAGATAGCACTGATGCTACTGGTACTGTTAGATTCCCTTTCACATTCAAtagatagatagtggccgcatttcgcagatgtTGTTGGTGGGTCTGCTGGTGTAGTTGATTTGCAAAACTCAAAATGTTTCACACCAGGTACTAATGAAGT' #@param {type: \"string\"}\n",
        "\n",
        "if sequence_length < sub_sequence_length:\n",
        "  sequence_length = sequence_length // 2\n",
        "\n",
        "if len(custom_sub_sequence) != 0:\n",
        "  subsequence = custom_sub_sequence\n",
        "else:\n",
        "  subsequence = str(target_gene_sequence.seq)[sub_sequence_start : sub_sequence_start + sub_sequence_length]\n",
        "\n",
        "from IPython.display import HTML, display\n",
        "\n",
        "def set_css():\n",
        "  display(HTML('''\n",
        "  <style>\n",
        "    pre {\n",
        "        white-space: pre-wrap;\n",
        "    }\n",
        "  </style>\n",
        "  '''))\n",
        "\n",
        "get_ipython().events.register('pre_run_cell', set_css)\n",
        "\n",
        "print(\"Subsequence length:  {}\".format(len(subsequence)))\n",
        "print(\"Subsequence: \\n{}\".format(subsequence))"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "Rn0n3r2rpOHI",
        "colab_type": "text"
      },
      "source": [
        "We'll use the [Eukaryotic Pathogen CRISPR guide RNA/DNA Design Tool](http://grna.ctegd.uga.edu/) to identify potential oligos for our CRISPR target.\n",
        "\n",
        "You need to provide a \"Job Name\" to EuPaGDT (the default name is the CPAR2 id followed by the size of the subsequence):"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "m040dBZKH9J7",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Specify your EuPaGDT job name:\n",
        "job_name = \"Als7-repair\" #@param {type:\"string\"}"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "rd4AfIrNH9wN",
        "colab_type": "text"
      },
      "source": [
        "The next step is to prepare values for the different fields that we have to submit when we request guide RNA:"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "ruqCpsAEGjii",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Fill in the relevant EuPaGDT form fields:\n",
        "rna_guided_nuclease_selection = \"SpCas9\"  #@param [\"SpCas9\"] {type: \"raw\"} #\n",
        "# unsupported options: \"SaCas9\", \"AsCpf1\", \"CjCas9\", \"CjCas9_2\", \"FnCas9\", \"FnCpf1\", \"LbCpf1\", \"NmCas9\", \"St1Cas9\", \"TdCas9\"]\n",
        "gRNAlength = 20  #@param {type:\"integer\"}\n",
        "PAM = \"NGG\"  #@param {type:\"string\"}\n",
        "homology_arm_length = 30  #@param {type:\"integer\"}\n",
        "HDR_repair_template_insertion = \"\"  #@param {type: \"string\"}\n",
        "genome = \"C_parapsilosis_CDC317_current_chromosomes\"  #@param {type:\"string\"}\n",
        "three_pam_pos = True\n",
        "# the subsequence to search was determined in a previous step\n",
        "\n",
        "form_fields = {\n",
        "    \"name\": (None, job_name),\n",
        "    \"proteinselection\": (None, rna_guided_nuclease_selection),\n",
        "    \"MH_switch\": (None, 'off'),  # Microhomology search\n",
        "    \"conreg_switch\": (None, 'off'),  # Conserved region search\n",
        "    \"textbox1\": (None, subsequence),\n",
        "    \"gRNAlength\": (\"ui-id-27\", gRNAlength),\n",
        "    \"PAM\": (\"ui-id-27\", PAM),\n",
        "    \"PAMpos\": (\"ui-id-27\", \"3PAM\" if three_pam_pos else \"5PAM\"),\n",
        "    \"onTflanklen\": (\"ui-id-29\", \"50\"),\n",
        "    \"onTidentity\": (\"ui-id-29\", \"0.7\"),\n",
        "    \"onTcoverage\": (\"ui-id-29\", \"0.7\"),\n",
        "    \"seedlength\": (\"ui-id-31\", \"15\"),\n",
        "    \"maxnummismatch\": (\"ui-id-31\", \"3\"),\n",
        "    \"offPAM\": (\"ui-id-31\", \"NAG,NGA\"),\n",
        "    \"HDRflank\": (\"ui-id-33\", homology_arm_length),\n",
        "    \"HDRinsertion\": (\"ui-id-33\", HDR_repair_template_insertion),\n",
        "    \"genome\": (\"ui-id-5\", genome),\n",
        "}\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "CBjNAQ9XfqkX",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Click the run icon to submit your request to find gRNA sequences from EuPaGDT\n",
        "# because the EuPaGDT site uses CGI scripts that are not open source,\n",
        "# and there is no evidence that it provides an API on its server,\n",
        "# we will have to emulate using a browser to make our requests.\n",
        "# using the Selenium tool to do this.\n",
        "\n",
        "# first, create some automation code that knows how to interact with the EuPaGDT page\n",
        "class EuPaGDTPage(object):\n",
        "  def __init__(self, driver, form_fields):\n",
        "    \"\"\"\n",
        "    takes a selenium webdriver instance, and a dictionary specifying the \n",
        "    field names and desired values in the page's form\n",
        "    \"\"\"\n",
        "    self.driver = driver\n",
        "    self.main_tab_handle = self.driver.window_handles[0]\n",
        "    self.form_fields = form_fields\n",
        "\n",
        "  def submit_request(self):\n",
        "      \"\"\" submit the request \"\"\"\n",
        "      for field_name in self.form_fields:\n",
        "          accordion_id, field_value = form_fields[field_name]\n",
        "          if accordion_id is not None:\n",
        "              accordion = self.driver.find_element_by_id(accordion_id)\n",
        "              controlled_div_id = accordion.get_attribute(\"aria-controls\")\n",
        "              controlled_div = self.driver.find_element_by_id(controlled_div_id)\n",
        "              if not controlled_div.is_displayed():\n",
        "                  accordion.click()\n",
        "\n",
        "          # Now try to find the field...\n",
        "          try:\n",
        "              field = self.driver.find_element_by_id(field_name)\n",
        "          except NoSuchElementException:\n",
        "              # some do not have the id attribute, so try looking by name\n",
        "              field = self.driver.find_element_by_name(field_name)\n",
        "\n",
        "          # once we've expanded enough accordions, we have to be sure to make the target field visible in the window\n",
        "          self.driver.execute_script(\"arguments[0].scrollIntoView();\", field)\n",
        "          if field.tag_name == \"select\":\n",
        "              Select(field).select_by_value(field_value)\n",
        "          elif field.tag_name == \"input\":\n",
        "              if field.get_attribute(\"type\") == \"radio\":\n",
        "                  radio_option = self.driver.find_element_by_css_selector(\n",
        "                      \"input[name='{}'][value='{}']\".format(field_name, field_value))\n",
        "                  self.driver.execute_script(\"arguments[0].scrollIntoView();\", radio_option)\n",
        "                  radio_option.click()\n",
        "              else:\n",
        "                  field.send_keys(field_value)\n",
        "          elif field.tag_name == \"textarea\":\n",
        "              field.send_keys(field_value)\n",
        "\n",
        "      form = self.driver.find_element_by_name(\"nameemailseq\")\n",
        "      form.submit()\n",
        "\n",
        "  def get_grna_table(self):\n",
        "    \"\"\" retrieve the html table with the results. \"\"\"\n",
        "    for window_handle in self.driver.window_handles:\n",
        "      if window_handle != self.main_tab_handle:\n",
        "        self.driver.switch_to.window(window_handle)\n",
        "\n",
        "    # confirm that we're in the correct tab for the run status:\n",
        "    if \"Design Tool\" in self.driver.title:\n",
        "      raise Exception(\"could not find the gRNA finder browser tab\")\n",
        "    report_handle = self.driver.current_window_handle\n",
        "\n",
        "    # the cgi is frequently reloading this page, which may result in a race condition where our target_links\n",
        "    # become stale.\n",
        "    # go find the target link...the one with the right text, but not the one for \"conserved\"\n",
        "    grna_finder_timeout = 4*60\n",
        "    target_link = [a for a in WebDriverWait(self.driver, grna_finder_timeout).until(\n",
        "      EC.presence_of_all_elements_located((By.LINK_TEXT, \"gRNA sequence and score\")),\n",
        "      \"gRNA finder browser page did not complete in {} seconds\".format(grna_finder_timeout))\n",
        "    if \"conserved\" not in a.get_attribute(\"href\")][0]\n",
        "\n",
        "    target_link.click()\n",
        "\n",
        "    # the tool is generating yet another tab\n",
        "    for window_handle in self.driver.window_handles:\n",
        "      if window_handle not in (self.main_tab_handle, report_handle):\n",
        "        self.driver.switch_to.window(window_handle)\n",
        "\n",
        "    # find the HTML table containing the results of the search\n",
        "    table = WebDriverWait(self.driver, 60).until(\n",
        "      EC.presence_of_element_located((By.CSS_SELECTOR, \"table\")))\n",
        "    # load it into a pandas dataframe for easier manipulation\n",
        "    data_frames = read_html(table.get_attribute(\"outerHTML\"), header=0)\n",
        "    df = data_frames[0]\n",
        "\n",
        "    # We are only interested in entries that have no off-target hits\n",
        "    # and on-target hits that are conditional on if we \n",
        "    # specified a custom_sub_sequence\n",
        "    if len(custom_sub_sequence) == 0:\n",
        "      on_target_hits = '1 | 0'\n",
        "    else:\n",
        "      on_target_hits = '0 | 0'\n",
        "    \n",
        "    f = df[(df['On-target hits in the genome (perfect-match | non-perfect-but-PAM-match)']==on_target_hits) & (df['Off-target hits (perfect-match | nonperfect-match)']=='0 | 0')]\n",
        "\n",
        "    # we've observed that we're still seeing duplicate entries of gRNA sequences, even with single on-target hits\n",
        "    # so we'll find any that are duplicated and ignore those\n",
        "    dupped = f.duplicated(subset=['gRNA sequence (PAM \"NGG\" )'], keep=False)\n",
        "    # report any that are not duplicated and sort tyem by the gRNA sequence\n",
        "    no_dups = f[~dupped]\n",
        "\n",
        "    # we need to remove any rows that have the potential problem of 4 or more T(U)s in a row\n",
        "    contains_4_or_more = no_dups['Potential problems during transcription'].str.contains('gRNA contains >=4')\n",
        "    no_or_limited_potential_problems = no_dups[~contains_4_or_more]\n",
        "\n",
        "    # now, we need to expand our set for any that require an A or a G prefix\n",
        "    rows_require_g_or_a = no_or_limited_potential_problems[\n",
        "      no_or_limited_potential_problems['Potential problems during transcription'].str.contains('please manually add a leading')]\n",
        "\n",
        "    # Do this by duplicating this set of rows twice, providing one of the prefixes to each set\n",
        "    rows_with_g_prefix = rows_require_g_or_a.copy()\n",
        "    rows_with_g_prefix['gRNA sequence (PAM \"NGG\" )'] = 'G' + rows_with_g_prefix['gRNA sequence (PAM \"NGG\" )']\n",
        "    rows_with_a_prefix = rows_require_g_or_a.copy()\n",
        "    rows_with_a_prefix['gRNA sequence (PAM \"NGG\" )'] = 'A' + rows_with_a_prefix['gRNA sequence (PAM \"NGG\" )']\n",
        "\n",
        "    # then removing the original rows and adding back the two new sets\n",
        "    no_problems = no_dups[no_dups['Potential problems during transcription'] == 'No problem found']\n",
        "    final_guide_rnas = no_problems.append(rows_with_g_prefix).append(rows_with_a_prefix).set_index('gRNA sequence (PAM \"NGG\" )')\n",
        "\n",
        "    return final_guide_rnas.sort_values(by=['gRNA id'])\n",
        "\n",
        "# set selenium webdriver options to be headless (so it won't bring up a browser)\n",
        "# assumes that the Chrome browser is installed.\n",
        "webdriver_options = webdriver.ChromeOptions()\n",
        "webdriver_options.add_argument('--headless')\n",
        "webdriver_options.add_argument('--no-sandbox')\n",
        "webdriver_options.add_argument('--disable-dev-shm-usage')\n",
        "\n",
        "# open the website, fill in the form, and wait for results\n",
        "driver = webdriver.Chrome('chromedriver',\n",
        "                          options=webdriver_options,\n",
        "                          desired_capabilities=None,\n",
        "                          service_log_path=None,\n",
        "                          chrome_options=None,\n",
        "                          keep_alive=True)\n",
        "driver.get(\"http://grna.ctegd.uga.edu/\")\n",
        "\n",
        "eupagdt = EuPaGDTPage(driver, form_fields)\n",
        "eupagdt.submit_request()\n",
        "eupagdt_results = eupagdt.get_grna_table()\n",
        "print(\"found {} gRNA sequences to explore\".format(eupagdt_results.shape[0]))\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "KBQzJBwBWYFu",
        "colab_type": "text"
      },
      "source": [
        "# Verify target insertion sites against RNA Fold Server"
      ]
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "dgj9QosE3qW3",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Click the run icon to pull RNA Fold information for the gRNA sequences\n",
        "# now update the table with information from the RNA Fold server about each guide RNA\n",
        "# first, create some code to define how to interact with the RNA fold server\n",
        "\n",
        "# Page object for the RNA Fold Server page\n",
        "class RNAFoldServerPage(object):\n",
        "  def __init__(self, driver):\n",
        "    \"\"\"\n",
        "    Takes an instance of a selenium webdriver\n",
        "    \"\"\"\n",
        "    self.driver = driver\n",
        "    # keep track of the browser tab where we start from\n",
        "    self.main_tab_handle = self.driver.window_handles[0]\n",
        "\n",
        "  def retrieve_fold_results(self, guide_rna_sequence):\n",
        "    \"\"\"\n",
        "    this method will populate the RNA Fold Server step 1 form\n",
        "    and request that our guide_rna_sequence be sent for processing,\n",
        "    then wait for the results before retrieving them.\n",
        "\n",
        "    ..note::\n",
        "    We are currently accepting all default form parameters, except for\n",
        "    the sequence that we are asking to be screened.\n",
        "\n",
        "    :param guide_rna_sequence: the gRNA sequence we need to check\n",
        "    \"\"\"\n",
        "    form_fields = {\n",
        "        \"SCREEN\": guide_rna_sequence,\n",
        "        # \"CONSTRAINT\": \"\",\n",
        "        # \"FILE\": \"\",\n",
        "        \"method\": \"p\",\n",
        "        \"nocloseGU\": False,\n",
        "        \"noLP\": True,\n",
        "        # \"dangling\": \"nd2\",\n",
        "        # \"param\": \"rna2004\",\n",
        "        # \"SHAPEDATA\": \"\",\n",
        "        # \"SHAPEFILE\": \"\",\n",
        "        # \"shapemethod\": \"deigan\",\n",
        "        # \"shape_slope\": \"1.9\",\n",
        "        # \"shape_intercept\": \"-0.7\",\n",
        "        # \"shape_beta\": \"0.8\",\n",
        "        # \"deigan_conversion\": \"linearlog\",\n",
        "        # \"shape_conv_cutoff\": \"0.25\",\n",
        "        # \"shape_conv_linear_s\": \"0.68\",\n",
        "        # \"shape_conv_linear_i\": \"0.2\",\n",
        "        # \"shape_conv_linearlog_s\": \"1.6\",\n",
        "        # \"shape_conv_linearlog_i\": \"-2.29\",\n",
        "        # \"Temp\": \"37\",\n",
        "        \"svg\": True,\n",
        "        \"reliability\": True,\n",
        "        \"mountain\": True,\n",
        "        # \"EMAIL\": \"\",\n",
        "    }\n",
        "\n",
        "    for field_name in form_fields:\n",
        "      # Now try to find the field...\n",
        "      try:\n",
        "        field = self.driver.find_element_by_id(field_name)\n",
        "      except NoSuchElementException:\n",
        "        # some do not have the id attribute, so try looking by name\n",
        "        field = self.driver.find_element_by_name(field_name)\n",
        "\n",
        "      # and set our desired value\n",
        "      self.driver.execute_script(\"arguments[0].scrollIntoView();\", field)\n",
        "      if field.tag_name == \"select\":\n",
        "        Select(field).select_by_value(form_fields[field_name])\n",
        "      elif field.tag_name == \"input\":\n",
        "        if field.get_attribute(\"type\") == \"radio\":\n",
        "          radio_option = self.driver.find_element_by_css_selector(\n",
        "            \"input[name='{}'][value='{}']\".format(field_name, form_fields[field_name]))\n",
        "          self.driver.execute_script(\"arguments[0].scrollIntoView();\", radio_option)\n",
        "          radio_option.click()\n",
        "        else:\n",
        "          field.send_keys(form_fields[field_name])\n",
        "      elif field.tag_name == \"textarea\":\n",
        "        field.send_keys(form_fields[field_name])\n",
        "\n",
        "    # submit the form for processing\n",
        "    form = self.driver.find_element_by_name(\"form\")\n",
        "    form.submit()\n",
        "\n",
        "    # wait for the page to refresh with our results...\n",
        "    # we'll use a simple way to determine that, look for a link that will only show once completed.\n",
        "    rna_fold_timeout = 4*60\n",
        "    WebDriverWait(self.driver, rna_fold_timeout).until(\n",
        "      EC.presence_of_all_elements_located((By.LINK_TEXT, \"color by base-pairing probability\")),\n",
        "      \"RNA fold server search request did not complete in {} seconds\".format(rna_fold_timeout))\n",
        "\n",
        "    # now go for the minimum free energy (MFE) numbers and the Centroid secondary structure svg diagram\n",
        "    # the mfe number in kcal/mol is at the following css path\n",
        "    mfe = self.driver.find_element_by_css_selector(\"p:nth-of-type(1) b\").text\n",
        "\n",
        "    # svg centroid secondary structure is inside an iframe look for the src containing \"centroid.svg\":\n",
        "    # <iframe src=\"http://rna.tbi.univie.ac.at/RNAfold/wjqi1FDuxB/test_sequenc_centroid.svg\" style=\"border:0;\" width=\"452\" height=\"650\">\n",
        "    iframes = self.driver.find_elements_by_tag_name('iframe')\n",
        "    centroid_iframe = [f for f in iframes if 'centroid.svg' in f.get_attribute('src')][0]\n",
        "\n",
        "    # return the MFE, and the link to the svg file displayed in the results\n",
        "    return mfe, centroid_iframe.get_attribute(\"src\")\n",
        "\n",
        "# create a new pandas DataFrame with as many rows as we have gRNAs, and columns\n",
        "# for the Minimum Free Energy and Centroid secondary structure SVG diagram\n",
        "rna_fold_mfe_svgs_df = DataFrame(index=eupagdt_results.index, columns=['MFE', 'Centroid secondary structure'])\n",
        "\n",
        "# parse through the guide RNAs to populate our new DataFrame\n",
        "n = 0\n",
        "for i in eupagdt_results.index:\n",
        "    # strip off the PAM [last three characters] when we pass in our guide RNA\n",
        "    driver.get(\"http://rna.tbi.univie.ac.at/cgi-bin/RNAWebSuite/RNAfold.cgi\")\n",
        "    rna_fold_server = RNAFoldServerPage(driver)\n",
        "\n",
        "    n += 1\n",
        "    s = i[:-3]\n",
        "    print(\"\\rretrieving {} of {} RNA Fold results for {}\".format(n, rna_fold_mfe_svgs_df.shape[0], s), end=\"\")\n",
        "\n",
        "    mfe, centroid_svg = rna_fold_server.retrieve_fold_results(s)\n",
        "\n",
        "    # add the results to our DataFrame\n",
        "    rna_fold_mfe_svgs_df.at[i, 'MFE'] = mfe\n",
        "    rna_fold_mfe_svgs_df.at[i, 'Centroid secondary structure'] = centroid_svg\n",
        "\n",
        "# merge the MFE and associated SVG diagram with our guide RNA results\n",
        "guide_rna_fold_mfe_svgs_df = concat([eupagdt_results, rna_fold_mfe_svgs_df], axis=1)\n",
        "\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "FZAxK3QeZOnr",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Display gRNA sequence scores with Minimum Free Energy and Centroid secondary structure\n",
        "display_filters = False  #@param {type: \"boolean\"}\n",
        "\n",
        "# present the data\n",
        "options.display.max_columns = None\n",
        "options.display.max_rows = guide_rna_fold_mfe_svgs_df.shape[0] + 1\n",
        "displayed_columns = [\"gRNA id\", \"Total score\", \"GC content\", \"efficiency score based on Doench et al.2014\", \"efficiency score based on CRISPRater\", \"MFE\", \"Centroid secondary structure\"]\n",
        "sort_by_columns = ['Total score', 'MFE']\n",
        "sort_order = [False, True]\n",
        "\n",
        "if display_filters:\n",
        "  # use google.colab.data_table to display the DataFrame with search/sort tools\n",
        "  %load_ext google.colab.data_table\n",
        "  display(guide_rna_fold_mfe_svgs_df[displayed_columns].sort_values(sort_by_columns, ascending=sort_order))\n",
        "else:\n",
        "  # display the DataFrame using standard HTML of the Colab notebook\n",
        "  # if we previously loaded the google.colab.data_table, unload it to avoid\n",
        "  # a possible error later\n",
        "  import subprocess\n",
        "  output = subprocess.getoutput('jupyter nbextension list')\n",
        "  if 'google.colab.data_table' in output:\n",
        "    %unload_ext google.colab.data_table\n",
        "\n",
        "  # provide some special formatting for the HTML results\n",
        "  def path_to_image_html(path):\n",
        "    # actually display the image that is referenced in the \"Centroid...\" column\n",
        "    return '<img src=\"'+ path + '\" width=\"60\">'\n",
        "\n",
        "  def highlight_gRNA(sequence):\n",
        "    # we want to highlight the PAM part of the gRNA sequence\n",
        "    pam_sequence = sequence[:-3] + '<span style=\"color: red;white-space: nowrap;\">{}</span>'.format(sequence[-3:])\n",
        "    final_sequence = pam_sequence\n",
        "    # we want to highlight the leading A or G if it was prefixed\n",
        "    if len(sequence) == gRNAlength + len(PAM) + 1:\n",
        "      final_sequence = '<span style=\"color: green;\">{}</span>'.format(pam_sequence[0]) + pam_sequence[1:]\n",
        "    return final_sequence\n",
        "\n",
        "  def highlight_mfe_0(mfe):\n",
        "    # highlight if the MFE is zero\n",
        "    if str(mfe) == \"0.00\":\n",
        "      mfe = '<span style=\"color: red\">{}</span>'.format(mfe)\n",
        "    return mfe\n",
        "\n",
        "  results_formatters = {\n",
        "      \"Centroid secondary structure\": path_to_image_html, \n",
        "      '__index__': highlight_gRNA,\n",
        "      \"MFE\": highlight_mfe_0\n",
        "      }\n",
        "\n",
        "  display(HTML(guide_rna_fold_mfe_svgs_df[displayed_columns].sort_values(sort_by_columns, ascending=sort_order).to_html(escape=False, formatters=results_formatters)))\n"
      ],
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "metadata": {
        "id": "WI-yKNvoA0dQ",
        "colab_type": "code",
        "cellView": "form",
        "colab": {}
      },
      "source": [
        "#@title Click Run to download the results table\n",
        "results_file_name = \"{}_gRNA_targets\".format(job_name)\n",
        "guide_rna_fold_mfe_svgs_df[displayed_columns].sort_values(sort_by_columns, ascending=sort_order).to_html(results_file_name + '.html', escape=False, formatters={\"Centroid secondary structure\": path_to_image_html})\n",
        "files.download(results_file_name + '.html')\n",
        "guide_rna_fold_mfe_svgs_df.sort_values(sort_by_columns, ascending=sort_order).to_csv(results_file_name + '.csv')\n",
        "files.download(results_file_name + '.csv')\n",
        "\n",
        "display(\"Look in your default Downloads folder for both a CSV and a HTML file with names of the pattern {}.*\".format(results_file_name))\n"
      ],
      "execution_count": null,
      "outputs": []
    }
  ]
}